Add additional sensors and settings to Roborock vacuums

starkillerOG opened the pull request:
Add additional sensors and improve the already implemented sensors for the Roborock S7 MaxV.

Change the decorator ids to `qualified name ` instead of simply name in order to prevent unique_id collisions in HomeAssistant for example with `DNDStatus.start` and `CleaningDetails.start`

Implement `choices_attribute` for setting decorator such that Enums can be evaluated at runtime by retrieving it from the device.

Split vacuum enums to a separte file to prevent circular imports

Controls added:
 - Mop routing (enum)
 - Mop scrub intensity (enum)
 - Mulit map (which floor is active, enum)
 - Auto dust collection (switch)
Buttons added:
 - start dust collection
 - stop dust collection
sensors added:
 - floor 0-4 clean start (when the last clean was per floor)
 - Last clean map name
 
![Roborock_s7MaxV - 1](https://user-images.githubusercontent.com/14811189/194712846-bcae24da-7263-4d67-b869-2c8fec79fd28.PNG)
![Roborock_s7MaxV - 2](https://user-images.githubusercontent.com/14811189/193644442-33fe3fc6-6ec1-4860-98bb-9f792f5630a3.PNG)
![Roborock_s7MaxV - 3](https://user-images.githubusercontent.com/14811189/193644446-a2311229-43f7-4b57-af60-d86f595e4c09.PNG)

starkillerOG commented:
@rytilahti I think the tests are now actually passing, but there is a problem with uploading to coverage.

starkillerOG commented:
@rytilahti would you please have a look at this PR?
Possibly even test is with your vacuum?
Would really like to get this merged in before the next python-miio release.

starkillerOG commented:
@hallister Thanks for the review, do you have any additional feedback?

rytilahti commented on miio/integrations/vacuum/roborock/vacuumcontainers.py:
```diff
@@ -353,6 +484,77 @@ def complete(self) -> bool:
+
+    @property
+    @sensor(
+        "Floor 0 clean start",
```
Instead of hardcoding the floor numbers, maybe just override the `sensors()` to return a generated list of sensor descriptors? This way it'll be easy to extend in case some models support different number of floors.

Alternatively, if other information besides the start date is interesting, the main container could embed "FloorDetails" container?

starkillerOG replied:
Alright, I made the `FloorCleanDetails` status container dynamic such that it will also work for any number of floors.
It does make it a bit hard to read, and unfortunatley Properties can not be set of an instance of a class, so I needed to also overwrite the __repr__.

The main status() already embeds the `FloorCleanDetails`.

starkillerOG replied:
I also embeded CleanDetails for each floor in the FloorCleanDetails container, so other details can also be accesed.

starkillerOG commented on miio/devicestatus.py:
```diff
@@ -100,7 +100,9 @@ def switches(self) -> Dict[str, SwitchDescriptor]:
-    def settings(self) -> Dict[str, SettingDescriptor]:
+    def settings(
+        self,
+    ) -> Dict[str, Union[EnumSettingDescriptor, NumberSettingDescriptor]]:
```
done

starkillerOG commented on miio/devicestatus.py:
```diff
@@ -254,13 +259,13 @@ def decorator_setting(func):
-                # TODO: adding choices from attribute is a bit more complex, as it requires a way to
-                # construct enums pointed by the attribute
-                raise NotImplementedError("choices_attribute is not yet implemented")
+            # if choices_attribute is not None:
```
done

starkillerOG commented on miio/device.py:
```diff
@@ -242,9 +243,13 @@ def buttons(self) -> List[ButtonDescriptor]:
-        settings = self.status().settings()
+        settings = (
+            self.status().settings()
+        )  # NOTE that this already does IO so schould be run in executer job in HA
```
I can move all the binding/IO (status retrieving, choices list retrieving etc) to the `info()` call, cache it there and then return the list of sensors/buttons/settings from the cache.
If sensors/buttons/settings is accesed when cache is still empty I can call `info()` there.

Do you want me to move it?

starkillerOG replied:
Or we could raise an error when sensors/buttons/settings is accesed when cache is still empty, maybe that is nicer since you then know that the sensors/buttons/settings will not do IO, and therefore do not have to be run in a executor job...

starkillerOG commented on miio/integrations/vacuum/roborock/vacuum.py:
```diff
@@ -403,9 +318,14 @@ def manual_control(
+        status.embed(clean_history)
+        clean_details = self.last_clean_all_floor(history=clean_history)
+        status.embed(clean_details["last"])
+        status.embed(FloorCleanDetails(clean_details))
```
I now return a tuple of (FloorCleanDetails, CleaningDetails) with the cleandetails the last clean.
They are combined in one call to reduce IO to the vacuum.

Yes the logic of which models support which properties schould defenitly be inside python-miio.
I was thinking about a @models decorator that could be added to a property or call to indicate which models do/do-not support a feature, if the decorator is not added it is assumed it is supported by all models, if the decorater is supplied the function will return None immediately withouth IO to the device if the model is in the do-not support list.
If a model is used that is not in the do/do-not support list a warning could be logged to request the user to report it to this library for his model.

starkillerOG commented on miio/integrations/vacuum/roborock/vacuum.py:
```diff
@@ -433,6 +353,44 @@ def map(self):
+    @command()
+    def get_multi_maps(self, skip_cache=False):
+        """Return list of multi maps."""
+        # {'max_multi_map': 4, 'max_bak_map': 1, 'multi_map_count': 3, 'map_info': [
```
I made a MultiMapList container

starkillerOG commented on miio/integrations/vacuum/roborock/vacuum.py:
```diff
@@ -519,18 +477,65 @@ def clean_history(self) -> CleaningSummary:
+            history = self.clean_history()
+
+        N_maps = self.get_multi_maps()["multi_map_count"]
+        map_ids = list(range(0, N_maps))
```
done

starkillerOG commented on miio/integrations/vacuum/roborock/vacuumcontainers.py:
```diff
@@ -86,15 +88,16 @@ def __init__(self, data: Dict[str, Any]) -> None:
         # 'water_shortage_status': 0, 'dock_type': 0, 'dust_collection_status': 0,
         # 'auto_dust_collection': 1,  'mop_mode': 300, 'debug_mode': 0}]
         self.data = data
+        self._multi_maps = multi_maps
```
I do not get what you mean.
The multi_map information is nessesary to map the current multi_map_id (e.g. 1) where the Roborock is now to a actual floor name (e.g. "Upstairs").

starkillerOG commented on miio/integrations/vacuum/roborock/vacuumcontainers.py:
```diff
@@ -206,15 +259,15 @@ def is_on(self) -> bool:
 
     @property
-    @sensor("Mop Attached")
+    @sensor("Mop attached")
```
I took this form the HomeAssistant entity naming standars:
See https://developers.home-assistant.io/blog/2022/07/10/entity_naming/

> Device, Area, and Entity names all start with a capital letter, the rest of the words are lower case (unless it's a word that represents a brand/name/abbreviation of course).

starkillerOG commented:
@rytilahti I finished processing your review comments, thanks a lot for the review!
I also implemented the button decorator and added two buttons to the roborock s7maxV.

Once you have some time, could you have a second go at a review?

craigcabrey commented:
This is great! Any idea if the clean & dirty tank status is exposed?

h3rbst commented:
This looks really great! Do you know if it would be possible to start the additional Dryer Module as well?
I am using the S7 Pro Ultra where the Dryer currently is not supported by the app. Would be nice if it is possible via Home Assistant.

rytilahti commented on miio/device.py:
```diff
@@ -238,13 +242,32 @@ def status(self) -> DeviceStatus:
         """Return device status."""
         raise NotImplementedError()
 
+    def cached_status(self) -> DeviceStatus:
```
Let's not add caching on this PR as it isn't directly relevant to the improvements for the roborock vacuums, and it will require some more investigation on what is the best approach to implement the functionality you are aiming to fix with this.

rytilahti commented on miio/devicestatus.py:
```diff
@@ -235,35 +230,32 @@ def setting(
                 name=name,
                 unit=unit,
-                setter=setter,
+                setter=None,
```
It feels like this changes the behavior and breaks cases where the `setter` is used (like upcoming generic miot integration I'm working on).

rytilahti commented on miio/integrations/vacuum/roborock/vacuumcontainers.py:
```diff
@@ -143,33 +202,51 @@ def error(self) -> str:
-        max_value=100,
-        step=1,
-        icon="mdi:fan",
-    )
```
Why remove this?

rytilahti commented on miio/integrations/vacuum/roborock/vacuumcontainers.py:
```diff
@@ -179,6 +256,28 @@ def map(self) -> bool:
         """Map token."""
         return bool(self.data["map_present"])
 
+    @property
+    @setting(
+        "Multi map",
+        choices_attribute="multi_map_enum",
+        setter_name="load_multi_map_by_enum",
+        icon="mdi:floor-plan",
+    )
+    def multi_map_id(self) -> int:
+        """The id of the current map with regards to the multi map feature,
+
+        [3,7,11,15] -> [0,1,2,3].
+        """
+        return int((self.data["map_status"] + 1) / 4 - 1)
+
+    @property
+    def multi_map_name(self) -> str:
+        """The name of the current map with regards to the multi map feature."""
+        if self._multi_maps is None:
+            return str(self.multi_map_id)
+
```
See my comment above wrt the multi maps. Let's try to keep this PR simple and add support for multi maps in a separate PR :-)

rytilahti commented on miio/integrations/vacuum/roborock/vacuumcontainers.py:
```diff
@@ -353,6 +534,58 @@ def complete(self) -> bool:
+    def __init__(self, data: Dict[str, Any]) -> None:
+        self.data = data
+
+        for map_id in self.data:
+            if self.data[map_id] is None:
+                setattr(self, f"CleanDetails_{map_id}", None)
+                setattr(self, f"start_{map_id}", None)
+                continue
+            setattr(self, f"CleanDetails_{map_id}", self.data[map_id])
+            setattr(self, f"start_{map_id}", self.data[map_id].start)
```
This looks rather messy and hard to follow through the setattr "magic", I think we can need to find a better solution for doing this. See my comment above wrt multiroom support.

rytilahti reviewed (changes requested):
Sorry for taking so long to respond, the travels took their toll but now I'm home again and have some free time to spare. Please split this PR into smaller ones: buttons, the multimap support, and caching should be discussed in separate PRs. This will both make it easier to review them and make the git history more readable for any future regression triaging.

starkillerOG commented:
> This is great! Any idea if the clean & dirty water tank status is exposed?

I did not see that in the status response, but I have a s7MaxV+ (auto-empty dock) not the Ultra with the Ultra dock, so I can't test the water tank status.
Do you see the clean&dirty tank status in the miio home app?

starkillerOG commented:
> This looks really great! Do you know if it would be possible to start the additional Dryer Module as well? I am using the S7 Pro Ultra where the Dryer currently is not supported by the app. Would be nice if it is possible via Home Assistant.

@h3rbst Xioami Miio is based on the protocol used by the xiaomi miio app, so if it is not available in the xiaomi miio app, it will not be possible to implement in HomeAssistant.
Starting a dust collection will be added as a button (but that is also possible in the xiaomi miio app).

crshray commented:
> > This is great! Any idea if the clean & dirty water tank status is exposed?
> 
> @craigcabrey I did not see that in the status response, but I have a s7MaxV+ (auto-empty dock) not the Ultra with the Ultra dock, so I can't test the water tank status. Do you see the clean&dirty tank status in the miio home app?

I tested with my s7 MaxV ultra and it looks like clean&dirty tank status is exposed via the `dock_error_status` field in the get_status response, if clean water tank is empty, `dock_error_status` will have a value of 38, if dirty water tank is full, `dock_error_status` will have a value of 39, if both errors happen at the same time, `dock_error_status` will have a value of 38 (i.e. clean water tank error takes precedence over the dirty water tank error).

craigcabrey commented:
> > This is great! Any idea if the clean & dirty water tank status is exposed?
> 
> @craigcabrey I did not see that in the status response, but I have a s7MaxV+ (auto-empty dock) not the Ultra with the Ultra dock, so I can't test the water tank status. Do you see the clean&dirty tank status in the miio home app?

Missed this, sorry! Yea as noted by @crshray it is available in the app.

starkillerOG commented:
@crshray thanks for the info, I added a dock_error and dock_error_code sensor in the latest commit.
Unfortunately I could not find a list of "dock_error_status" codes, so I only added 0, 38 and 39 for now.

starkillerOG commented:
@h3rbst yes that is certainly possible, but I would need to have decoded packet captures to figure out the commands that you would like to have added to HomeAssistant (preferablly just capture as much as possible).
See the discussion here: https://github.com/home-assistant/core/issues/69869

starkillerOG commented:
Now that the last piece of this PR has been split out here: https://github.com/rytilahti/python-miio/pull/1614
This PR can be closed.

h3rbst commented:
@starkillerOG is it possible to activate this for the S7 Pro Ultra as well? The S7 Pro Ultra has the same Dock as the MaxV

starkillerOG commented:
@h3rbst These features are added for all models that support them.
Although it is still in dev and not released yet.
We first need a new python-miio release.

h3rbst commented:
> @h3rbst These features are added for all models that support them. Although it is still in dev and not released yet. We first need a new python-miio release.

Ah okay, I see. I thought, since the PR is already merged, that it is part of the new December release.

xsasx commented:
@starkillerOG Sorry for maybe my stupidness - as i understand you rigt - i should have with the normal miio integration in homeassistant the feature that you created with the MOP intensity etc ? Cause i have it not with my Roborock S7 or do i need to change any file under my Homeassistant server with your files (replace)

starkillerOG commented:
@xsasx No that needs a new version of Python-miio to be released.
I have been waiting on a new version for a long time...

Maybe @rytilahti could shed some light on when a new version would be expected.

neforce commented:
i'm experiencing sometimes trouble loading the map of the Xiaomi app. However, roborock app itself is working will. But when i use roborock app, i miss a lot of sensors which i'm using now. If i read this thread well, is it true that i can use roborock app on my phone, and use the roborock integration of Home Assistant after this commit in this issue has been merged?

JR2012thman commented:
> @xsasx No that needs a new version of Python-miio to be released. I have been waiting on a new version for a long time...
> 
> Maybe @rytilahti could shed some light on when a new version would be expected.

Is there any movement on this? I am happy to send commands as needed but this would be a great set of features for those who are less technical.
